File: 97/solution.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int s1Len = s1.length();
        int s2Len = s2.length();
        
        if (s1Len + s2Len != s3.length()) return false;
        
        vector<vector<bool>> dp(s1Len + 1, vector<bool>(s2Len + 1, false));
        
        dp[0][0] = true;
        for (int i=0; i<=s1Len; i++) {
            for (int j=0; j<=s2Len; j++) {
                if (i || j) {
                    bool isIMeet = (i && dp[i-1][j] && s1[i-1] == s3[i+j-1]);
                    bool isJMeet = (j && dp[i][j-1] && s2[j-1] == s3[i+j-1]);
                    dp[i][j] = isIMeet || isJMeet;
                }
            }
        }
        
        return dp[s1Len][s2Len];
    }
};

File: 97/solution.go
func isInterleave(s1 string, s2 string, s3 string) bool {
	s1Len := len(s1)
	s2Len := len(s2)
	
	if s1Len + s2Len != len(s3) {
			return false
	}
	
	dp := make([][] bool, s1Len + 1)
	for i := range(dp) {
			dp[i] = make([] bool, s2Len + 1)
	}
	
	dp[0][0] = true
	
	for i := range(dp) {
			for j := range(dp[i]) {
					if i > 0 || j > 0 {
							isIMeet := i > 0 && dp[i-1][j] && s1[i-1] == s3[i+j-1]
							isJMeet := j > 0 && dp[i][j-1] && s2[j-1] == s3[i+j-1]
							dp[i][j] = isIMeet || isJMeet
					}
			}
	}
	return dp[s1Len][s2Len]
	
}